### PR TITLE
Feature landing view

### PR DESCRIPTION
**Pull request by tmafaralala**
Collection View implementation with music player per cell

### CODE REVIEW

File: free-tunes/Controllers/TrackCollectionViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//
+//  Created by Tshwarelo Mafaralala on 2022/03/13.
+//
+
+import UIKit
+
+class TrackCollectionViewCell: UICollectionViewCell {
+
+//  MARK: - Interface Builder Outlets
+    @IBOutlet weak var trackCover: UIImageView!
```
Comment (kieranwoodrow) [inline free-tunes/Controllers/TrackCollectionViewCell.swift]: Maybe make these variables private

**Review by kieranwoodrow — DISMISSED** [review]

File: free-tunes/Controllers/TrackCollectionViewCell.swift (hunk 1 above)
Comment (tmafaralala) [reply free-tunes/Controllers/TrackCollectionViewCell.swift]: but that entails that the variables can't be accessed when setting up the cell in the  collection view

Comment (smabhena) [reply free-tunes/Controllers/TrackCollectionViewCell.swift]: You can use setters for those variables

**Review by smabhena — DISMISSED** [review]
LGTM 👍

File: free-tunes/Controllers/OnBoardingViewController.swift [hunk 2]
```diff
@@ -9,7 +9,10 @@ import UIKit
 
 class OnBoardingViewController: UIViewController {
 
-    @IBOutlet weak var headerText: UILabel!
+// MARK: - Interface Builder Outlets
+    @IBOutlet private weak var headerText: UILabel!
```
Comment (KirstyPallatte) [inline free-tunes/Controllers/OnBoardingViewController.swift:13]: As a suggestion I would Make every IBOutlet and IBAction private

File: free-tunes/Controllers/TrackExplorerViewController.swift [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    
+// MARK: - Controller Logic Methods
+    func setUpExploreCollectionView() {
+        trackCollectionView.dataSource = self
+        trackCollectionView.delegate = self
+    }
+    
+    func getExploreViewContent() {
+        let url = Constants.artistExploreContentUrl
```
Comment (KirstyPallatte) [inline free-tunes/Controllers/TrackExplorerViewController.swift:40]: Should this not be guard  let url = Constants.artistExploreContentUrl see { return } as the url may not exist so it should be guarded

File: free-tunes/Controllers/OnBoardingViewController.swift (hunk 2 above)
Comment (tmafaralala) [reply free-tunes/Controllers/OnBoardingViewController.swift:13]: I believe it is the case unless if i'm missing something

File: free-tunes/Controllers/TrackExplorerViewController.swift (hunk 3 above)
Comment (tmafaralala) [reply free-tunes/Controllers/TrackExplorerViewController.swift:40]: the case where the url doesn't exist is handled in the make request method already

Comment (KirstyPallatte) [reply free-tunes/Controllers/TrackExplorerViewController.swift:40]: That makes sense. Thank you Noah for the clarification. Happy to approve now

**Review by KirstyPallatte — DISMISSED** [review]
Looks good to me. Good job!

File: free-tunes/Controllers/TrackExplorerViewController.swift [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import UIKit
+import AVFoundation
+
+class TrackExplorerViewController: UIViewController {
+    
+// MARK: - Interface Builder Outlets
+    @IBOutlet private weak var trackCollectionView: UICollectionView!
+    
+    private lazy var exploreViewModel = TrackExplorerViewModel()
+    var looper: AVPlayerLooper!
```
Comment (bgouws) [inline free-tunes/Controllers/TrackExplorerViewController.swift]: These can also be private 😄

**Review by bgouws — CHANGES_REQUESTED** [review]
Just a little bit of clean up 😄

**Comment by bgouws** [comment]
Would also be nice to see a screenshot or two 😄

File: free-tunes/Controllers/TrackExplorerViewController.swift (hunk 4 above)
Comment (tmafaralala) [reply free-tunes/Controllers/TrackExplorerViewController.swift]: missed i'll check it out

**Review by KirstyPallatte — APPROVED** [review]

**Review by KirstyPallatte — DISMISSED** [review]
Looks good to me 👍🏻

File: free-tunes/Controllers/LoginViewController.swift [hunk 5]
```diff
@@ -39,20 +50,7 @@ class LoginViewController: UIViewController {
...
-    func checkCredentials(username: String?, password: String?) -> Bool {
-        if username != "Admin" || password != "TestPass123" {
-            return false
-        }
-        return true
-    }
-
+// MARK: - Input Validation
     func checkTextFieldData() -> Bool {
         if usernameInput.text != "" && passwordInput.text != "" {
```
Comment (lhendricks-DVT) [inline free-tunes/Controllers/LoginViewController.swift]: Have you looked into isEmpty?

File: free-tunes/Controllers/TrackExplorerViewController.swift [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        trackCollectionView.register(UINib(nibName: "TrackCollectionViewCell", bundle: .main), forCellWithReuseIdentifier: "TrackCollectionViewCell")
+        trackCollectionView.isPagingEnabled = true
+        getExploreViewContent()
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        self.navigationController?.setNavigationBarHidden(true, animated: animated)
+        super.viewWillAppear(animated)
```
Comment (lhendricks-DVT) [inline free-tunes/Controllers/TrackExplorerViewController.swift:29]: Your super should always be the first thing in a method

File: free-tunes/Controllers/TrackExplorerViewController.swift [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        if player.volume != 0 {
+            player.volume = 0
+            return
+        }
+        player.volume = 1.0
+    }
+    
+// MARK: - Collection View Custom Methods
+    func setupTrackCell(cell: TrackCollectionViewCell?, index: Int ) {
```
Comment (lhendricks-DVT) [inline free-tunes/Controllers/TrackExplorerViewController.swift]: Think about moving this function to be apart of your cell TrackCollectionViewCell

**Review by lhendricks-DVT — CHANGES_REQUESTED** [review]

File: free-tunes/Controllers/LoginViewController.swift (hunk 5 above)
Comment (tmafaralala) [reply free-tunes/Controllers/LoginViewController.swift]: I never thought of it actually. Let me have  a look at it👍

File: free-tunes/Controllers/TrackExplorerViewController.swift (hunk 6 above)
Comment (tmafaralala) [reply free-tunes/Controllers/TrackExplorerViewController.swift:29]: I had the changes locally, I must have lost them. I'll update it.

File: free-tunes/Controllers/TrackExplorerViewController.swift (hunk 7 above)
Comment (tmafaralala) [reply free-tunes/Controllers/TrackExplorerViewController.swift]: Valid point, I'll update it

**Review by simekanidev — DISMISSED** [review]

File: free-tunes/Controllers/TrackExplorerViewController.swift (hunk 6 above)
Comment (bgouws) [reply free-tunes/Controllers/TrackExplorerViewController.swift:29]: Looks like this still needs to be addressed 😄

**Review by bgouws — CHANGES_REQUESTED** [review]

**Comment by lhendricks-DVT** [comment]
Just a general note rather create more frequent PRs to mitigate large files numbers. Clients dont like anything more than 10-12 files being reviewed in a pull request

**Comment by tmafaralala** [comment]
> Just a general note rather create more frequent PRs to mitigate large files numbers. Clients dont like anything more than 10-12 files being reviewed in a pull request

noted

File: free-tunes/Controllers/TrackExplorerViewController.swift (hunk 6 above)
Comment (tmafaralala) [reply free-tunes/Controllers/TrackExplorerViewController.swift:29]: I keep trying to update it but it's just not outdating the code

**Review by KirstyPallatte — DISMISSED** [review]
Looks good to me 👍🏻

**Review by smabhena — CHANGES_REQUESTED** [review]
Note that there's merge conflicts

**Review by smabhena — APPROVED** [review]

**Review by KirstyPallatte — APPROVED** [review]
Looks good 👍🏻

**Review by lhendricks-DVT — APPROVED** [review]

**Review by bgouws — APPROVED** [review]

**Pull request merged**